File: src/lib/fuzzy.ts
/**
 * Subsequence matcher tuned for short identifier-ish strings (model names,
 * command names, file paths). Returns a score plus the matched indices so the
 * caller can highlight; `null` means no match.
 */

export interface FuzzyMatch {
  score: number
  indices: number[]
}

export function fuzzy(needle: string, haystack: string): FuzzyMatch | null {
  if (!needle) return { score: 0, indices: [] }
  const query = needle.toLowerCase()
  const target = haystack.toLowerCase()

  // Fast path: a contiguous hit always beats a scattered one.
  const direct = target.indexOf(query)
  if (direct >= 0) {
    const indices = Array.from({ length: query.length }, (_, i) => direct + i)
    const boundary = direct === 0 || /[^a-z0-9]/.test(target[direct - 1] ?? "")
    return { score: 1000 - direct + (boundary ? 220 : 0), indices }
  }

  const indices: number[] = []
  let score = 0
  let cursor = 0
  let streak = 0
  for (const char of query) {
    const at = target.indexOf(char, cursor)
    if (at < 0) return null
    const boundary = at === 0 || /[^a-z0-9]/.test(target[at - 1] ?? "")
    streak = at === cursor ? streak + 1 : 0
    score += 12 + streak * 8 + (boundary ? 26 : 0) - Math.min(at - cursor, 12)
    indices.push(at)
    cursor = at + 1
  }
  return { score, indices }
}

export function rank<T>(items: T[], query: string, key: (item: T) => string): T[] {
  if (!query.trim()) return items
  const scored: Array<{ item: T; score: number }> = []
  for (const item of items) {
    const match = fuzzy(query.trim(), key(item))
    if (match) scored.push({ item, score: match.score })
  }
  scored.sort((a, b) => b.score - a.score)
  return scored.map((entry) => entry.item)
}

/** Split a string into matched / unmatched runs for highlight rendering. */
export function segments(text: string, indices: number[]) {
  if (!indices.length) return [{ text, hit: false }]
  const set = new Set(indices)
  const out: Array<{ text: string; hit: boolean }> = []
  let buffer = ""
  let hit = set.has(0)
  for (let i = 0; i < text.length; i += 1) {
    const isHit = set.has(i)
    if (isHit !== hit) {
      if (buffer) out.push({ text: buffer, hit })
      buffer = ""
      hit = isHit
    }
    buffer += text[i]
  }
  if (buffer) out.push({ text: buffer, hit })
  return out
}

File: desk-rs/src/fuzzy.rs
//! Subsequence matching, ported from the web build so both front ends rank
//! the same query the same way.
//!
//! Tuned for short identifier-ish strings — session titles, model names, file
//! paths. The shape that matters: a contiguous hit always outranks a scattered
//! one, and a hit that starts at a word boundary outranks one that starts
//! mid-word. Without those two rules a subsequence matcher happily puts
//! "s-e-s-s-i-o-n" spread across a paragraph above the session actually called
//! "session".

/// Score `needle` against `haystack`, or `None` when it does not match at all.
pub fn score(needle: &str, haystack: &str) -> Option<i32> {
    if needle.is_empty() {
        return Some(0);
    }
    let query: Vec<char> = needle.to_lowercase().chars().collect();
    let target: Vec<char> = haystack.to_lowercase().chars().collect();

    // Fast path: a contiguous run beats anything scattered, so it is worth
    // looking for one before walking character by character.
    if let Some(at) = contains_at(&target, &query) {
        let boundary = at == 0 || !target[at - 1].is_alphanumeric();
        return Some(1000 - at as i32 + if boundary { 220 } else { 0 });
    }

    let mut total = 0i32;
    let mut cursor = 0usize;
    let mut streak = 0i32;
    for wanted in query {
        let at = target[cursor..].iter().position(|c| *c == wanted)? + cursor;
        let boundary = at == 0 || !target[at - 1].is_alphanumeric();
        // A run of adjacent matches compounds: "gpui" hitting four consecutive
        // characters should not score the same as four scattered ones.
        streak = if at == cursor { streak + 1 } else { 0 };
        total += 12 + streak * 8 + if boundary { 26 } else { 0 }
            - (at - cursor).min(12) as i32;
        cursor = at + 1;
    }
    Some(total)
}

/// Index of the first contiguous occurrence of `needle` in `target`.
fn contains_at(target: &[char], needle: &[char]) -> Option<usize> {
    if needle.len() > target.len() {
        return None;
    }
    (0..=target.len() - needle.len()).find(|&start| target[start..start + needle.len()] == *needle)
}
